core/app/components/pubs/PubEditor/PubEditorClient.tsx

```diff
@@ -56,28 +61,48 @@ const preparePayload = ({
 		HydratedRelatedFieldValue | SingleFormValues | undefined | StagesId
 	>;
 	arrayDefaults: Record<string, HydratedRelatedFieldValue>;
+	contextEditorGetters: ContextEditorGetters;
 }) => {
 	const valuesPayload: Record<string, HydratedRelatedFieldValue[] | JsonValue | Date> = {};
-	for (const { slug } of formElements) {
-		if (
-			slug &&
-			toggleContext.isEnabled(slug) &&
-			// Only send fields that were changed. RHF erroneously reports fields as changed (dirty)
-			// sometimes even when the default value is the same, so we do the check ourselves
-			!isEqualWith(formValues[slug], defaultValues[slug])
-		) {
-			const val = formValues[slug];
-			if (Array.isArray(val)) {
-				const filteredVal = val.filter((v: RelatedFieldValue) => {
-					const isNew = !v.valueId;
-					const isChanged = v.valueId && !isEqualWith(arrayDefaults[v.valueId], v);
-					return isNew || isChanged;
-				});
-				valuesPayload[slug] = filteredVal;
-			} else {
-				valuesPayload[slug] = val;
+	for (const { slug, schemaName } of formElements) {
+		if (!slug) {
+			continue;
+		}
+
+		if (!toggleContext.isEnabled(slug)) {
+			continue;
+		}
+
+		if (isEqualWith(formValues[slug], defaultValues[slug])) {
+			continue;
+		}
+
+		const val = formValues[slug];
+
+		// we manually retrieve the value from the context editor here instead of having RHF set it
+		// bc this is much faster for very large documents
+		if (schemaName === CoreSchemaType.RichText && val === EvilContextEditorSymbol) {
+			const getter = contextEditorGetters[slug];
+			if (!getter) {
+				throw new Error(`No getter found for slug ${slug}`);
 			}
+			const value = getter.current?.getCurrentState();
+
+			valuesPayload[slug] = value && serializeProseMirrorDoc(value.doc);
```

Comment:
we could also just convert it to HTML here, let me know your opinion on that! that way, the server action does not need to be able to handle prosemirror at all, and we can maybe simplify the validation code somewhat